app/services/sign_in/refresh_token_decryptor.rb

```diff
@@ -0,0 +1,59 @@
+# frozen_string_literal: true
+
+module SignIn
+  class RefreshTokenDecryptor
+    attr_reader :encrypted_refresh_token
+
+    def initialize(encrypted_refresh_token:)
+      @encrypted_refresh_token = encrypted_refresh_token
+    end
+
+    def perform
+      split_refresh_token = split_encrypted_refresh_token(encrypted_refresh_token)
+
+      decrypted_part = get_decrypted_part(split_refresh_token)
+      nonce = get_nonce_part(split_refresh_token)
+      version = get_version_part(split_refresh_token)
+
+      validate_token!(decrypted_part, nonce, version)
+
+      SignIn::RefreshToken.new(session_handle: decrypted_part.session_handle,
+                               user_uuid: decrypted_part.user_uuid,
+                               parent_refresh_token_hash: decrypted_part.parent_refresh_token_hash,
+                               anti_csrf_token: decrypted_part.anti_csrf_token,
+                               nonce: decrypted_part.nonce,
+                               version: decrypted_part.version)
+    end
+
+    private
+
+    def validate_token!(decrypted_token, nonce, version)
+      raise SignIn::Errors::RefreshVersionMismatchError if decrypted_token.version != version
+      raise SignIn::Errors::RefreshNonceMismatchError if decrypted_token.nonce != nonce
+    end
+
+    def get_decrypted_part(split_refresh_token)
+      decrypt_refresh_token(split_refresh_token[SignIn::Constants::REFRESH_TOKEN_ENCRYPTED_POSITION])
+    end
+
+    def get_nonce_part(split_refresh_token)
+      split_refresh_token[SignIn::Constants::REFRESH_TOKEN_NONCE_POSITION]
+    end
+
+    def get_version_part(split_refresh_token)
+      split_refresh_token[SignIn::Constants::REFRESH_TOKEN_VERSION_POSITION]
```

Comment:
Could you describe the benefit of making these consants?

For example, these methods could be removed in favor of assuming their position:

```ruby
token, nonce, version = encrypted_refresh_token.split '.'
```

And likewise for building:

```ruby
encrypted_refresh_token = [token, nonce, version].join '.'
```

Reply:
I made these constants because the positioning of the components is kinda subtle, and since there are two objects that have to separately infer the locations of these components, I was trying to store that in a shared location (array locations as constants).

For example, if we decide that we want to change the order of these components or add a field for a V1 version of this, we just have to change the constants, or we can have constants that describe the makeup of a V0 vs V1 refresh token, instead of that info being hidden in the encryptor/decryptor logic